src/server/bg_services/replication_log_parser.js

```diff
@@ -0,0 +1,320 @@
+        // then no need to process anything
+        if (!next_log_entry.Contents || next_log_entry.Contents.length === 0) {
+            dbg.log0('log_parser: no more log entries');
+            break;
+        }
+
+        const next_log_data = await aws_get_next_log(s3, logs_bucket, next_log_entry.Contents[0].Key);
+        aws_parse_log_object(candidates_obj, next_log_data, ignore_fn);
+
+        logs_retrieved_count -= 1;
+    }
+    while ((Object.keys(candidates_obj).length < candidates_limit) && logs_retrieved_count !== 0 && continuation_token);
+
+    return {
+        items: candidates_obj,
+        done: async () => {
+            if (continuation_token) {
+                await replication_store.update_log_replication_marker_by_id(replication_config._id, rule_id, { continuation_token });
+            }
+        },
+    };
+}
+
+async function aws_get_next_log_entry(s3, logs_bucket, logs_prefix, continuation_token) {
+    let start_after = logs_prefix;
+    if (start_after && !start_after.endsWith('/')) {
+        start_after += '/';
+    }
+
+    try {
+        dbg.log1('log_parser aws_get_next_log_entry: params:', logs_bucket, logs_prefix, continuation_token);
+        const res = await s3.listObjectsV2({
+            Bucket: logs_bucket,
+            Prefix: logs_prefix,
+            ContinuationToken: continuation_token,
+            MaxKeys: 1,
+            StartAfter: start_after,
+        }).promise();
+
+        dbg.log1('log_parser aws_get_next_log_entry: finished successfully ', res);
+        return res;
+
+    } catch (err) {
+        dbg.error('log_parser aws_get_next_log_entry: error:', err);
+        throw err;
+    }
+}
+
+async function aws_get_next_log(s3, bucket, key) {
+    try {
+        dbg.log1('log_parser aws_get_next_log: params:', bucket, key);
+        const res = await s3.getObject({
+            Bucket: bucket,
+            Key: key,
+            ResponseContentType: 'json'
+        }).promise();
+
+        dbg.log1('log_parser aws_get_next_log: finished successfully ', res);
+        return res;
+
+    } catch (err) {
+        dbg.error('log_parser aws_get_next_log: error:', err);
+        throw err;
+    }
+}
+
+function aws_parse_log_object(candidates_map, log_object, ignore_fn) {
+    const log_string = log_object.Body.toString();
+    const log_array = log_string.split("\n");
+
+    for (const line of log_array) {
+        if (line !== '') {
+            let log = parse_aws_log_entry(line);
+            if (log.operation) {
+                // ignore the log entry if it should be ignored
+                if (ignore_fn(log)) continue;
+
+                if (log.operation.includes('PUT.OBJECT') || log.operation.includes('POST.OBJECT')) {
+                    candidates_map[log.key] = {
+                        action: 'copy',
+                        time: log.time,
+                    };
+                }
+                if (log.operation.includes('DELETE.OBJECT') && log.http_status === 204) {
```

Comment:
is there a valid situation in which the heep_status will be other than 204 and the op is delete? it's not handled according to this code

Reply:
Actually the idea was to not skip every such operation which is of type `DELETE` but the HTTP_STATUS is 204. For some reason, I have seen multiple such operations which fail with `400` errors or with `MALFORMEDHEADER` error code. So the idea is to process only those operations which resulted in a successful deletions and ignore the rest.

Reply:
I meant to deletion with success http codes like 202, 204, 200, https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE
not sure if we would ever get those but please check

Reply:
interesting, so we won't support processing delete multiple objects event on this drop? probably we would like at least document it as a gap

Reply:
@romayalon, I performed `aws s3api delete-objects` on the bucket and finally received the logs for them. It turns out that the call results in `BATCH.DELETE.OBJECT` and a successful delete is indicated by `204`.

Attaching the relevant log here:
```
2e055fd7dc2c9eb01fe8d089e32b84c4b9e8a66b85e370793b4c7259ea477902 utk-repl [28/Feb/2023:14:15:42 +0000] 202.134.149.165 arn:aws:iam::255735097469:user/utkarsh H1SZ73V4C2X0DJXA BATCH.DELETE.OBJECT awsdata1 - 204 - - 10240 - - - - - BVQb2i2NLQHD1BcYkQZU58Ev1Elsz8+sQtOaYY9JU02SGdkuiQPHyXWcIuzUT/hC3YMS/LFNwRQ= SigV4 ECDHE-RSA-AES128-GCM-SHA256 AuthHeader utk-repl.s3.ap-south-1.amazonaws.com TLSv1.2 - -
2e055fd7dc2c9eb01fe8d089e32b84c4b9e8a66b85e370793b4c7259ea477902 utk-repl [28/Feb/2023:14:15:42 +0000] 202.134.149.165 arn:aws:iam::255735097469:user/utkarsh H1SZ73V4C2X0DJXA BATCH.DELETE.OBJECT awsdata2 - 204 - - 10240 - - - - - BVQb2i2NLQHD1BcYkQZU58Ev1Elsz8+sQtOaYY9JU02SGdkuiQPHyXWcIuzUT/hC3YMS/LFNwRQ= SigV4 ECDHE-RSA-AES128-GCM-SHA256 AuthHeader utk-repl.s3.ap-south-1.amazonaws.com TLSv1.2 - -
2e055fd7dc2c9eb01fe8d089e32b84c4b9e8a66b85e370793b4c7259ea477902 utk-repl [28/Feb/2023:14:15:42 +0000] 202.134.149.165 arn:aws:iam::255735097469:user/utkarsh H1SZ73V4C2X0DJXA REST.POST.MULTI_OBJECT_DELETE - "POST /?delete HTTP/1.1" 200 - 192 - 42 - "-" "aws-cli/2.10.2 Python/3.11.2 Darwin/22.3.0 source/arm64 prompt/off command/s3api.delete-objects" - BVQb2i2NLQHD1BcYkQZU58Ev1Elsz8+sQtOaYY9JU02SGdkuiQPHyXWcIuzUT/hC3YMS/LFNwRQ= SigV4 ECDHE-RSA-AES128-GCM-SHA256 AuthHeader utk-repl.s3.ap-south-1.amazonaws.com TLSv1.2 - -
```

Here I deleted 2 keys with delete-objects (the keys being `awsdata1` and `awsdata2`).

I think with this we have all of our bases covered.

Can I mark this as resolved now?

Reply:
great! if AWS docs say that it will return 204 on success, sure